### PR TITLE
feat(sgx): implement dynamic thread allocation

### PR DESCRIPTION
**Pull request by haraldh**
The SGX keep allocates memory and sets up the TCS for new threads, if the existing thread pool is drained.

The thread pool is not shrinked though. See https://github.com/enarx/enarx/issues/2251

Fixes: https://github.com/enarx/enarx/issues/2160

### CODE REVIEW

File: crates/shim-sgx/src/handler/mod.rs [hunk 1]
```diff
@@ -845,4 +879,78 @@ impl<'a> Handler<'a> {
...
+        // # Safety
+        // ThreadMem is a POD type and the memory is aligned and zeroed.
+        let tm = unsafe { &mut *(addr.as_ptr() as *mut ThreadMem) };
+
+        let stack_end_1 = NonNull::new(&mut tm.cssa_stack as *mut _ as *mut c_void).unwrap();
+        let stack_end_0 = NonNull::new(&mut tm.stack as *mut _ as *mut c_void).unwrap();
+        let tcs = &tm.tcs as *const Tcs;
+        let ssa = tm.ssa.as_ptr();
+
+        tm.tcs.ossa = ssa as u64 - shim_address() as u64;
```
Comment (bstrie) [inline crates/shim-sgx/src/handler/mod.rs:905]: Can we get some explanatory comments throughout this function to help the reader follow along?

File: tests/crates/enarx_exec_tests/src/bin/thread-many.rs [hunk 2]
```diff
@@ -2,23 +2,37 @@
...
-        let ret: i32 = thread2.join().unwrap();
-        assert_eq!(ret, 0);
+fn main() {
+    for i in 0..100 {
+        GLOBAL_THREAD_COUNT.fetch_add(1, Ordering::SeqCst);
+
+        thread::spawn(move || {
+            // do some work
+            eprintln!("{}-th thread reporting", i + 1);
+            thread::sleep(Duration::from_secs(1));
```
Comment (bstrie) [inline tests/crates/enarx_exec_tests/src/bin/thread-many.rs:20]: Does this increase the time it takes to run the tests compared to today?

Comment (haraldh) [reply tests/crates/enarx_exec_tests/src/bin/thread-many.rs:20]: old:
```
$ time cargo test --features dbg --test integration exec::thread_many -- --exact --nocapture 
[…]
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 43 filtered out; finished in 3.81s


real	0m4.075s
user	0m3.739s
sys	0m1.985s
```

new:
```
$ time cargo test --features dbg --test integration exec::thread_many -- --exact --nocapture 
[…]
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 43 filtered out; finished in 3.79s


real	0m4.034s
user	0m3.510s
sys	0m1.709s
```

File: crates/shim-sgx/src/handler/mod.rs (hunk 1 above)
Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:905]: yes, will do

Comment (haraldh) [reply crates/shim-sgx/src/handler/mod.rs:940]: done

File: src/backend/sgx/thread.rs [hunk 3]
```diff
@@ -161,6 +158,11 @@ impl super::super::Thread for Thread {
             _ => unreachable!(),
         }
 
+        if self.cssa > 4 {
```
Comment (bstrie) [inline src/backend/sgx/thread.rs:161]: The commit message for this says:
```
 fix(backend/sgx): exit for CSSA > 3

Do not enter a CSSA > 3, but exit the whole process.
```
What explains the discrepancy between `> 4` in the code and `> 3` in the commit message?

File: src/backend/sgx/enarxcall.rs [hunk 4]
```diff
@@ -24,6 +21,7 @@ use sallyport::item::enarxcall::sgx::{Report, TargetInfo};
 use sallyport::item::{enarxcall, Item};
 use tracing::{error, trace_span};
 
+// #[instrument(skip(keep, data), level = "trace")]
```
Comment (bstrie) [inline src/backend/sgx/enarxcall.rs:24]: Should this comment be left in?

File: src/backend/sgx/enarxcall.rs [hunk 5]
```diff
@@ -240,6 +245,7 @@ pub(crate) fn sgx_enarxcall<'a>(
                     *ret = (-err.err.raw_os_error().unwrap()) as usize;
                 }
             };
+            drop(fd_locked);
```
Comment (bstrie) [inline src/backend/sgx/enarxcall.rs:248]: Is this manual drop necessary? The guard is about to go out of scope anyway.

File: src/backend/sgx/ioctls.rs [hunk 6]
```diff
@@ -152,6 +156,45 @@ impl RestrictPermissions {
...
+                result: 0,
+                count: 0,
+            };
+
+            let fd_locked = mutex_fd.lock().unwrap();
+            let mut fd_cloned = fd_locked.try_clone().unwrap();
+            match ENCLAVE_RESTRICT_PERMISSIONS.ioctl(&mut fd_cloned, &mut parameters) {
+                Ok(_) => {}
+                // EINTR
+                Err(e) if matches!(e.raw_os_error(), Some(4)) => {
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:176]: Should this magic number be a constant from somewhere?

File: src/backend/sgx/ioctls.rs [hunk 7]
```diff
@@ -152,6 +156,45 @@ impl RestrictPermissions {
...
+                    debug!("ENCLAVE_RESTRICT_PERMISSIONS failed with EINTR");
+                }
+                // EWOULDBLOCK, EAGAIN
+                Err(e) if matches!(e.kind(), io::ErrorKind::WouldBlock) => {
+                    debug!("ENCLAVE_RESTRICT_PERMISSIONS failed with EAGAIN");
+                }
+                // EBUSY
+                Err(e) if matches!(e.raw_os_error(), Some(16)) => {
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:184]: Constant?

File: src/backend/sgx/ioctls.rs [hunk 8]
```diff
@@ -152,6 +156,45 @@ impl RestrictPermissions {
...
+                    debug!("ENCLAVE_RESTRICT_PERMISSIONS failed with EBUSY");
+                    std::thread::sleep(time::Duration::from_millis(1));
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+            drop(fd_locked);
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:192]: Necessary manual drop?

File: src/backend/sgx/ioctls.rs [hunk 9]
```diff
@@ -182,6 +225,45 @@ impl ModifyTypes {
...
+                result: 0,
+                count: 0,
+            };
+
+            let fd_locked = mutex_fd.lock().unwrap();
+            let mut fd_cloned = fd_locked.try_clone().unwrap();
+            match ENCLAVE_MODIFY_TYPES.ioctl(&mut fd_cloned, &mut parameters) {
+                Ok(_) => {}
+                // EINTR
+                Err(e) if matches!(e.raw_os_error(), Some(4)) => {
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:244]: Constant?

File: src/backend/sgx/ioctls.rs [hunk 10]
```diff
@@ -182,6 +225,45 @@ impl ModifyTypes {
...
+                    debug!("ENCLAVE_MODIFY_TYPES failed with EINTR");
+                }
+                // EWOULDBLOCK, EAGAIN
+                Err(e) if matches!(e.kind(), io::ErrorKind::WouldBlock) => {
+                    debug!("ENCLAVE_MODIFY_TYPES failed with EAGAIN");
+                }
+                // EBUSY
+                Err(e) if matches!(e.raw_os_error(), Some(16)) => {
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:252]: Constant?

File: src/backend/sgx/ioctls.rs [hunk 11]
```diff
@@ -182,6 +225,45 @@ impl ModifyTypes {
...
+                    debug!("ENCLAVE_MODIFY_TYPES failed with EBUSY");
+                    std::thread::sleep(time::Duration::from_millis(1));
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+            drop(fd_locked);
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:260]: Manual drop?

File: src/backend/sgx/ioctls.rs [hunk 12]
```diff
@@ -219,6 +301,43 @@ impl RemovePages {
...
+                length: self.length - count,
+                count: 0,
+            };
+
+            let fd_locked = mutex_fd.lock().unwrap();
+            let mut fd_cloned = fd_locked.try_clone().unwrap();
+            match ENCLAVE_REMOVE_PAGES.ioctl(&mut fd_cloned, &mut parameters) {
+                Ok(_) => {}
+                // EINTR
+                Err(e) if matches!(e.raw_os_error(), Some(4)) => {
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:319]: Constant?

File: src/backend/sgx/ioctls.rs [hunk 13]
```diff
@@ -219,6 +301,43 @@ impl RemovePages {
...
+                    debug!("ENCLAVE_REMOVE_PAGES failed with EINTR");
+                }
+                // EWOULDBLOCK, EAGAIN
+                Err(e) if matches!(e.kind(), io::ErrorKind::WouldBlock) => {
+                    debug!("ENCLAVE_REMOVE_PAGES failed with EAGAIN");
+                }
+                // EBUSY
+                Err(e) if matches!(e.raw_os_error(), Some(16)) => {
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:327]: Constant?

File: src/backend/sgx/ioctls.rs [hunk 14]
```diff
@@ -219,6 +301,43 @@ impl RemovePages {
...
+                    debug!("ENCLAVE_REMOVE_PAGES failed with EBUSY");
+                    std::thread::sleep(time::Duration::from_millis(1));
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+            drop(fd_locked);
```
Comment (bstrie) [inline src/backend/sgx/ioctls.rs:335]: Manual drop?

File: src/backend/sgx/thread.rs (hunk 3 above)
Comment (haraldh) [reply src/backend/sgx/thread.rs:161]: oops

File: src/backend/sgx/ioctls.rs (hunk 6 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs:176]: right.. I could get them from `libc`

File: src/backend/sgx/thread.rs (hunk 3 above)
Comment (haraldh) [reply src/backend/sgx/thread.rs]: fixed

File: src/backend/sgx/enarxcall.rs (hunk 4 above)
Comment (haraldh) [reply src/backend/sgx/enarxcall.rs]: removed

File: src/backend/sgx/ioctls.rs (hunk 6 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: replaced

File: src/backend/sgx/ioctls.rs (hunk 9 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: replaced

File: src/backend/sgx/ioctls.rs (hunk 12 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: replaced

File: src/backend/sgx/ioctls.rs (hunk 13 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: replaced

File: src/backend/sgx/ioctls.rs (hunk 14 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: removed

File: src/backend/sgx/ioctls.rs (hunk 11 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: removed

File: src/backend/sgx/ioctls.rs (hunk 10 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: replaced

File: src/backend/sgx/ioctls.rs (hunk 8 above)
Comment (haraldh) [reply src/backend/sgx/ioctls.rs]: removed

**Review by bstrie — APPROVED** [review]

**Pull request merged**